### ISSUE TITLE
Suggestion: move "Model" methods to gl.Mesh

### ISSUE THREAD

**Issue by rafcolson** [GoranM/bdx#680]
I think it's a bit confusing to be having `Scene.models`, `GameObject.modelName()`, `GameObject.useUniqueModel()`, `GameObject.replaceModel()` along with `GameObject.mesh`, `GameObject.joinedMeshObjects` and `GameObject.updateJoinedMesh`. Especially when a Mesh is supposed to wrap the Model.

So I would like to suggest `Scene.meshes`, `Mesh.name()`, `Mesh.unique()` and `Mesh.replace()`.

**Comment by SolarLune** [comment]
That all sounds fine to me.

I think currently, BDX creates a new Mesh to wrap the Model and provide editing functions whenever necessary, but it doesn't use that Mesh when drawing objects and stuff. Shouldn't the Mesh have a ModelInstance, or be the means through which BDX renders GameObjects? It seems kinda weird that users use GameObject.mesh to edit the model, and BDX uses GameObject.modelInstance to render it. 

So basically, maybe we should move GameObject.modelInstance to the Mesh class for BDX to use?

**Comment by rafcolson** [comment]
I'm not sure. We couldn't have `Scene.meshes` contain references to the modelInstance, then could we? Because that wouldn't make much sense. For which case I still think it had been better if the mesh would be part of the modelInstance. And not having Mesh wrap the model, but the be an extension of LibGdx's Mesh. I pictured this kind of hierarchy:

GameObject -> Meshes (ext. ModelInstance / replace mesh method here - set(index, Mesh)) -> Mesh (ext. LibGdx's Mesh)

**Comment by SolarLune** [comment]
I believe our Mesh class contains a reference to the Model, so I don't quite get how GameObject.meshes should extend LibGDX's ModelInstance.

> We couldn't have Scene.meshes contain references to the modelInstance, then could we? 

Well, scene.meshes could contain references to BDX Meshes, each of which would have the stuff to render with (including a reference to the correct Model).

_____

But never mind; I guess it doesn't really matter as long as it's clear and consistent. For now, just moving the functions into the Mesh class would be fine. We can refactor later if it doesn't work or make sense.

**Comment by GoranM** [comment]
Sounds fine.

> Shouldn't the Mesh have a ModelInstance, or be the means through which BDX renders GameObjects? It seems kinda weird that users use GameObject.mesh to edit the model, and BDX uses GameObject.modelInstance to render it.

I think we discussed that in #610.

**Comment by SolarLune** [comment]
I was looking to add in #780, but before I do, it's probably wise to solve this first, and then I'll do both.
_____

I like the names @rafcolson suggested, so I'm good on those.

> For which case I still think it had been better if the mesh would be part of the modelInstance. And not having Mesh wrap the model, but the be an extension of LibGdx's Mesh. I pictured this kind of hierarchy:

> GameObject -> Meshes (ext. ModelInstance / replace mesh method here - set(index, Mesh)) -> Mesh (ext. LibGdx's Mesh)

I'm not quite seeing how this hierarchy would work - each GameObject only has one mesh, right? That's the mesh / model that's getting replaced when you use `replaceModel()`?

**Comment by rafcolson** [comment]
>I'm not quite seeing how this hierarchy would work - each GameObject only has one mesh, right? That's the mesh / model that's getting replaced when you use `replaceModel()`?

Only if multiple meshes per object are supported, a `Meshes` class would be useful. But it's not supported so, ...

**Comment by SolarLune** [comment]
OK, so maybe later we can refactor it to add multiple meshes.

So I was thinking of removing `replaceModel()` in favor of just having two different functions to replace the visual and physical aspects of a GameObject. Something like

```java

mesh.visual("Player 2");  // Replaces visual mesh
mesh.physical("LargerCollisionBox");  // Replaces collision mesh
mesh.physical();  // Updates collision mesh, if vertices are moved
```

Thoughts?

**Comment by GoranM** [comment]
In thinking about this further, it all seems kinda wrong:

First, I don't think it makes sense to have methods on the mesh that require access to the existing GameObject. A mesh should be a self contained resource, and if anyone wants to replace the existing mesh on a GameObject, they should be able to do it via a simple setter method.

We shouldn't conflate the mesh concept with the physics body concept. If you want to update the body's collision hull based on some mesh, there should be a method on the GameObject (or, on a class that would wrap/extend the bullet body) that would do that.

The mesh itself should have no knowledge of the physics body.

**Comment by SolarLune** [comment]
Those points make sense. 

I think the core problem is that the Mesh doesn't really represent the visual representation of the GameObject enough, but rather, is just the gateway through which changes (using the vertex functions) happen. We should probably, in addition to some function moving as mentioned before, make it so that this is the case. That would mean that one could do something like:

```java
Mesh treeMesh = scene.objects.get("Tree").mesh();
GameObject rock = scene.objects.get("Rock");
rock.mesh(treeMesh);  // Set the rock to use the tree's mesh
```

This should probably be the case for BDX internal, as well (so we would set the mesh on GameObjects using the `mesh()` functions, rather than setting the modelInstance and creating and setting a new Mesh manually).

In addition to the other refactoring, the `Mesh.unique()` function should simply duplicate the mesh, to be swapped in later using the `GameObject.mesh()` function, manually.

Finally, I'm still in favor of separated display and physical mesh replacement functions, and one to update the physics mesh based on an input mesh, as well.

**Comment by rafcolson** [comment]
For updating the body we could have a method: `GameObject.updateBody ()`. I agree that `Mesh` shouldn't be concerned about physics.

**Comment by GoranM** [comment]
Yeah, that would be better.

> In addition to the other refactoring, the Mesh.unique() function should simply duplicate the mesh, to be swapped in later using the GameObject.mesh() function, manually.

That would require the mesh to have a reference to the GameObject, and we're trying to avoid that.

If a GameObject needs to have it's own exclusive mesh, it should be possible to just make a copy: `gobj.mesh(gobj.mesh().copy());`.

**Comment by SolarLune** [comment]
> That would require the mesh to have a reference to the GameObject, and we're trying to avoid that.

How so? it would just return a duplicate copy of the Mesh, for the GameObject to apply with its own `GameObject.mesh()` function. It's basically the same as what you described above.

**Comment by SolarLune** [comment]
OK, so I made some progress on this. Some questions that came up:

1) How do we know when to dispose of unique meshes? Only when using GameObjects are ended?

2) It's kinda weird that materials are present on the GameObject and modified by the Mesh - shouldn't it be that the Mesh has its set of Materials (so that a Tree mesh looks like a tree, for example)?

**Comment by SolarLune** [comment]
OK, so I tested it out, and the materials being present on the Mesh doesn't really work, because it's assigning the materials of the model and not of the modelInstance. So, I guess I could just pass the modelInstance to the Mesh instead of the model itself to work with.

**Comment by SolarLune** [comment]
OK, so that was stupid, haha. Each Mesh needs to extend Model at most. I thought that `modelInstance` referred to the same nodes available in the Model - the ModelInstance is just the visual instance of the Model, so each Mesh can't have a single ModelInstance. I'm going to revert it back to put the Materials on the GameObject.

**Comment by GoranM** [comment]
> it would just return a duplicate copy of the Mesh

Shouldn't Mesh.copy do that?

Having Mesh.unique as just an alias for Mesh.copy doesn't really make sense, so that's why I assumed that it would call something like `gobj.mesh(this.copy())` internally, and would therefore require the gobj reference.

> How do we know when to dispose of unique meshes? Only when using GameObjects are ended?

Basically, yes. When there are no more game objects that are using a unique mesh, that mesh can be disposed.

> It's kinda weird that materials are present on the GameObject and modified by the Mesh - shouldn't it be that the Mesh has its set of Materials (so that a Tree mesh looks like a tree, for example)?

Yes, that would make more sense.

**Comment by SolarLune** [comment]
> Shouldn't Mesh.copy do that?

Calling it that is fine, I was just saying that functionally, my suggested function was the same as yours.

> When there are no more game objects that are using a unique mesh, that mesh can be disposed.

Hmm, I don't think that can be said definitively, since the user could theoretically be holding onto a reference. Perhaps we could dispose of a unique mesh only if it is active on a GameObject when it's end()-ed?

> Yes, that would make more sense.

I don't think that'll be possible as the Mesh would have to have access to ModelInstances to set their materials.

**Comment by GoranM** [comment]
> Calling it that is fine

The term "unique" couldn't really serve as a viable alternative, because it doesn't actually describe the operation in question.

> Hmm, I don't think that can be said definitively, since the user could theoretically be holding onto a reference. Perhaps we could dispose of a unique mesh only if it is active on a GameObject when it's end()-ed?

I assumed that the check would be performed on a used mesh, when a given GameObject was removed from the scene, so yes, that was the idea.

It still doesn't prevent the user from holding on to a mesh reference for which the model was disposed, but you have to draw the line somewhere.

I guess we could have an "invalid" method, like we do for GameObject ... Or, assuming that Java has a "destructor" mechanism like python's `__del__` method, maybe we could use that?

It seems like it would be worth looking into.

> I don't think that'll be possible as the Mesh would have to have access to ModelInstances to set their materials.

Wouldn't you have to create a new ModelInstance whenever you set the mesh anyway? The mesh could have a method for creating the model instance, and it could set the materials at that point, just before it returns.

Also, assuming that materials on the model instance are full copies, rather than references to originals hosted on the mesh, it wouldn't be a big problem to actually hold a list of model instances on the mesh, which the mesh could update whenever mesh materials are changed.

That seems like it could work fairly well, if you wanted to centralize graphical resources on the Mesh object (which seems like it would be cleaner, conceptually, if not implementation-wise).

**Comment by SolarLune** [comment]
>  Or, assuming that Java has a "destructor" mechanism like python's __del__ method, maybe we could use that?

It doesn't seem like it does - I read that it's because it's a garbage collected language, so you don't know just when it'll be collected. I'm guessing that's why LibGDX has `destroy()` functions for memory-heavy classes (like image-related classes).

> Wouldn't you have to create a new ModelInstance whenever you set the mesh anyway? The mesh could have a method for creating the model instance, and it could set the materials at that point, just before it returns.

That could work.

> Also, assuming that materials on the model instance are full copies, rather than references to originals hosted on the mesh, it wouldn't be a big problem to actually hold a list of model instances on the mesh, which the mesh could update whenever mesh materials are changed.

I don't know, that seems kinda complex just to do what we're doing right now. While the end result is cleaner conceptually, the implementation does seem quite a bit more complex (?), and to the end user, the difference is kind of minimal (accessing materials just goes through the mesh rather than the GameObject). It does make a bit more sense, though.

Branch updated.

**Comment by GoranM** [comment]
> It doesn't seem like it does - I read that it's because it's a garbage collected language, so you don't know just when it'll be collected. I'm guessing that's why LibGDX has destroy() functions for memory-heavy classes (like image-related classes).

Java has a `finalize` method that seems to be synonymous with Python's `__del__`. It's called just before the object would be naturally disposed by the GC (which implies that there are no more valid references to that object).

I think certain libgdx objects are set to be manually disposable because releasing their memory can be fairly expensive, so there may be a need to directly control when that happens, in order to mitigate potential performance problems.

For us, the only thing that really matters is the mechanism by which we can know that there are no more valid references to a Mesh, and then we can either dispose the Model in finalize, or we can simply use finalize to place the Model in a "toBeDisposed" queue of some sort, which can then be handled by the engine.

> While the end result is cleaner conceptually, the implementation does seem quite a bit more complex

It would be a bit more code, but I doubt that it would be significantly more complex than what we have now.

It might not be a bad idea to prioritize conceptual clarity in this case.

**Comment by SolarLune** [comment]
> For us, the only thing that really matters is the mechanism by which we can know that there are no more valid references to a Mesh, and then we can either dispose the Model in finalize, or we can simply use finalize to place the Model in a "toBeDisposed" queue of some sort, which can then be handled by the engine.

Hmm, OK. I guess we can try it out and revert the change if anything goes wrong down the road.

> It would be a bit more code, but I doubt that it would be significantly more complex than what we have now.

OK, let me backtrack here. The problem I ran into earlier was that the Mesh class can't have its own set of materials, because then all GameObjects that use that mesh couldn't swap out or alter their materials without making a whole separate Mesh copy, or altering the Mesh for all other GameObjects (which isn't ideal). So, for example, you couldn't do `mesh().materials.get(0).color(...)`, because that affects the materials for the Mesh, not the materials for the modelInstance.

**Comment by GoranM** [comment]
> I guess we can try it out and revert the change if anything goes wrong down the road.

It should be possible to simply test if the system works: Is the dispose method being called? Is the memory actually freed?

These seem like questions that we should be able to answer before committing anything to master.

> The problem I ran into earlier was that the Mesh class can't have its own set of materials, because then all GameObjects that use that mesh couldn't swap out or alter their materials without making a whole separate Mesh copy, or altering the Mesh for all other GameObjects

I don't really see why it's a "problem". I mean, blender meshes seem to work in the same way: Objects use a given mesh, and that mesh is using a specific set of materials. If you want to use the same mesh, but with different materials, you create a new mesh with those materials.

**Comment by SolarLune** [comment]
> It should be possible to simply test if the system works: Is the dispose method being called? Is the memory actually freed?

Yeah, I tested it and it works, but I imagine problems could still arise depending on real-world usage due to issues we haven't foreseen. Dealing with memory allocation / deallocation under Java seems to be a bit fiddly at times.

> I don't really see why it's a "problem". I mean, blender meshes seem to work in the same way: Objects use a given mesh, and that mesh is using a specific set of materials. If you want to use the same mesh, but with different materials, you create a new mesh with those materials.

Hmm, that is the way Blender handles it - I'd forgotten. That'd work, I guess.
_____

OK, I think I got it. PR submitted.

**Comment by rafcolson** [comment]
I think I can close this now. Thanks @SolarLune.

**Comment by SolarLune** [comment]
No problem.
